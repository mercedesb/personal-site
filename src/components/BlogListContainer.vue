<template>
  <div class="PageContent PageContent--wide PageContent--fullWidth">
      <h3 class="Blog-activeCategory" v-if="activeCategory">
        Posts about {{this.activeCategory}}
      </h3>
    <div class="FlexContainer PageContent PageContent--wide">
      <BlogList :posts="posts" :color="color" :isFirstPage="this.currentPageNumber === 1" :loading="loading" />
      <div>
        <BlogCategories :categories="allCategories" :activeCategory="activeCategory" />
      </div>
    </div>
    <div class='Blog-pageNumberList'>
      <div v-if="!activeCategory">
        <span v-for="n in maxPageNumber"
          :key="n"
          :class="`Blog-pageNumber ${n == currentPageNumber ? 'Blog-pageNumber--current' : ''}`"
          v-on:click="getPage(n)"
        >
          {{n}}
        </span>
      </div>
    </div>
  </div>
</template>

<script>
import BlogList from './BlogList.vue'
import BlogCategories from './BlogCategories.vue'

export default {
  components: {
    BlogList, BlogCategories
  },
  props: {
    color: String,
    page: {
      type: Number,
      default: 1
    }
  },
  data () {
    return {
      currentPageNumber: this.page,
      pageSize: 10,
      filterByCategories: [],
      loading: false
    }
  },
  computed: {
    posts () {
      return this.$store.state.blogPosts.map((post) => {
        return {
          id: post.id,
          color: this.color,
          title: post.title,
          preamble: post.preamble,
          mainContent: post.mainContent,
          urlSegment: post.urlSegment,
          tags: post.tags,
          date: new Date(post.publishDate)
        }
      })
    },
    maxPageNumber () {
      return this.activeCategory ?  Math.ceil(this.posts.length / this.pageSize) : this.$store.state.maxBlogPostPages
    },
    allCategories () {
      return this.$store.state.blogCategories.map(c => { return { name: c } } ).sort(this.sortCategoryNames);
    },
    activeCategory () {
      const queryCategory = this.$route.query.filter ? this.$route.query.filter.toLowerCase() : ""
      const withIncorrectCasingHandled = this.allCategories.map(c => c.name).find(name => queryCategory === name.toLowerCase())
      return withIncorrectCasingHandled
    }
  },
  methods: {
    getPage (pageNumber) {
      this.loading = true;
      let query = { page: pageNumber, pageSize: this.pageSize }
      if (this.activeCategory) {
        query.filter = this.activeCategory
      }
      this.$store.dispatch('getBlogPosts', query).then(() => {
        this.loading = false;
      });
      this.currentPageNumber = pageNumber
    },
    sortCategoryNames (tagA, tagB) {
      let aIsLessThanB = tagA.name.toLowerCase() < tagB.name.toLowerCase();
      let aIsGreaterThanB = tagA.name.toLowerCase() > tagB.name.toLowerCase();

      if (aIsGreaterThanB) return 1
      if (aIsLessThanB) return -1
      else return 0
    }
  }
}
</script>

<style lang="scss">
@import '../assets/styles/variables.scss';

.Blog {
  &-activeCategory {
    margin: 0 $large-spacing $large-spacing;
    font-size: 1rem;
    font-weight: $heavy-font-weight;
  }

  &-pageNumberList {
    text-align: center;
    padding: $large-spacing 0;
  }

  &-pageNumber {
    margin: 0 $base-spacing;
    font-family: $header-font-family;
    font-size: $large-font-size;
    cursor: pointer;

    &--current {
      text-decoration: underline;
    }
  }
}
</style>
