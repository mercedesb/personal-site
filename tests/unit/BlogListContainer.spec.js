import BlogListContainer from "@/components/BlogListContainer.vue";
import { TestUtility } from "./TestUtility";

describe("BlogListContainer", () => {
  let component;

  const initialProps = {
    color: "gray"
  };

  const store = {
    state: {
      blogPosts: TestUtility.blogPosts,
      blogCategories: ["Test", "Categories", "for", "Specs"]
    }
  };

  const shallow = (propsData, localConfig) =>
    TestUtility.shallow(BlogListContainer, {
      store,
      propsData: {
        ...initialProps,
        ...propsData
      },
      ...localConfig
    });

  describe("Snapshots", () => {
    describe("with a color passed in", () => {
      it("matches snapshot", () => {
        component = shallow();
        expect(component.element).toMatchSnapshot();
      });
    });

    describe("without a color passed in", () => {
      it("matches snapshot", () => {
        component = shallow({ color: undefined });
        expect(component.element).toMatchSnapshot();
      });
    });

    describe("when its the first page", () => {
      it("matches snapshot", () => {
        component = shallow({ page: 1 });
        expect(component.element).toMatchSnapshot();
      });
    });

    describe("when its not the first page", () => {
      it("matches snapshot", () => {
        component = shallow({ page: 2 });
        expect(component.element).toMatchSnapshot();
      });
    });
  });

  describe("Computed", () => {
    describe("posts", () => {
      beforeEach(() => {
        component = shallow();
      });

      it("has posts populated", () => {
        expect(component.vm.posts).toBeDefined();
      });

      it("creates a color prop on the posts", () => {
        expect(component.vm.posts[0].color).toEqual(initialProps.color);
      });

      it("creates a date prop on the posts", () => {
        expect(component.vm.posts[0].date).toBeDefined();
      });
    });

    describe("allCategories", () => {
      beforeEach(() => {
        component = shallow();
      });

      it("has allCategories populated", () => {
        expect(component.vm.allCategories).toBeDefined();
      });

      it("creates a name prop on the posts", () => {
        expect(component.vm.allCategories[0].name).toBeDefined();
      });

      it("sorts the categories, case insensitive", () => {
        expect(component.vm.allCategories[0].name).toEqual("Categories");
        expect(component.vm.allCategories[1].name).toEqual("for");
        expect(component.vm.allCategories[2].name).toEqual("Specs");
        expect(component.vm.allCategories[3].name).toEqual("Test");
      });
    });
  });

  describe("Methods", () => {
    describe("getPage", () => {
      it("dispatches getBlogPosts to the store", () => {
        component = shallow();
        component.vm.getPage(2);
        expect(component.vm.$store.dispatch).toHaveBeenCalledWith(
          "getBlogPosts",
          { page: 2, pageSize: expect.anything() }
        );
      });
    });

    describe("sortCategoryNames", () => {
      it("sorts category object alphabetically by name, case insensitive", () => {
        let categories = [
          { name: "test" },
          { name: "The" },
          { name: "alpha" },
          { name: "Sorting" },
          { name: "By" },
          { name: "Name" }
        ];
        let expected = [
          { name: "alpha" },
          { name: "By" },
          { name: "Name" },
          { name: "Sorting" },
          { name: "test" },
          { name: "The" }
        ];
        component = shallow();
        expect(categories.sort(component.vm.sortCategoryNames)).toEqual(
          expected
        );
      });
    });
  });
});
