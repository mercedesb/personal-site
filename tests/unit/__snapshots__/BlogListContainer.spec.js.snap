// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogListContainer Snapshots when its not the first page matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide PageContent--fullWidth"
>
  <!---->
   
  <div
    class="FlexContainer PageContent PageContent--wide"
  >
    <bloglist-stub
      color="gray"
      posts="[object Object],[object Object],[object Object]"
    />
     
    <div>
      <blogcategories-stub
        categories="[object Object],[object Object],[object Object],[object Object]"
        color="purple"
      />
    </div>
  </div>
   
  <div
    class="Blog-pageNumberList"
  >
    <div />
  </div>
</div>
`;

exports[`BlogListContainer Snapshots when its the first page matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide PageContent--fullWidth"
>
  <!---->
   
  <div
    class="FlexContainer PageContent PageContent--wide"
  >
    <bloglist-stub
      color="gray"
      isfirstpage="true"
      posts="[object Object],[object Object],[object Object]"
    />
     
    <div>
      <blogcategories-stub
        categories="[object Object],[object Object],[object Object],[object Object]"
        color="purple"
      />
    </div>
  </div>
   
  <div
    class="Blog-pageNumberList"
  >
    <div />
  </div>
</div>
`;

exports[`BlogListContainer Snapshots with a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide PageContent--fullWidth"
>
  <!---->
   
  <div
    class="FlexContainer PageContent PageContent--wide"
  >
    <bloglist-stub
      color="gray"
      isfirstpage="true"
      posts="[object Object],[object Object],[object Object]"
    />
     
    <div>
      <blogcategories-stub
        categories="[object Object],[object Object],[object Object],[object Object]"
        color="purple"
      />
    </div>
  </div>
   
  <div
    class="Blog-pageNumberList"
  >
    <div />
  </div>
</div>
`;

exports[`BlogListContainer Snapshots without a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide PageContent--fullWidth"
>
  <!---->
   
  <div
    class="FlexContainer PageContent PageContent--wide"
  >
    <bloglist-stub
      isfirstpage="true"
      posts="[object Object],[object Object],[object Object]"
    />
     
    <div>
      <blogcategories-stub
        categories="[object Object],[object Object],[object Object],[object Object]"
        color="purple"
      />
    </div>
  </div>
   
  <div
    class="Blog-pageNumberList"
  >
    <div />
  </div>
</div>
`;
